src/fdTransfer.cpp

```diff
@@ -0,0 +1,416 @@
+
+// base header for all requests
+enum request_type {
+    PERF_FD,
+    KALLSYMS_FD,
+};
+struct fd_request {
+    enum request_type type;
+    unsigned int length;
+#define MAX_REQUEST_LENGTH 2048
+    unsigned int request_id;
+};
+
+struct perf_fd_request {
+    struct fd_request header;
+    pid_t tid;
+    struct perf_event_attr attr;
+};
+
+#define TMP_KALLSYMS_PATH "/tmp/async-profiler-kallsyms.%d"
+
+int FdTransfer::_listener = -1;
+int FdTransfer::_peer = -1;
+unsigned int FdTransfer::_request_id = 0;
+
+static bool send_all(int fd, const void *buf, size_t count, bool *eof) {
+    const unsigned char *p = (const unsigned char*)buf;
+    size_t n = 0;
+
+    while (n < count) {
+        ssize_t sent = send(fd, p + n, count - n, 0);
+        if (sent <= 0) {
+            *eof == sent == 0;
+            return false;
+        }
+        n += sent;
+    }
+
+    return true;
+}
+
+static bool recv_all(int fd, void *buf, size_t count, bool *eof) {
+    unsigned char *p = (unsigned char*)buf;
+    size_t n = 0;
+
+    while (n < count) {
+        ssize_t recvd = recv(fd, p + n, count - n, 0);
+        if (recvd <= 0) {
+            *eof = recvd == 0;
+            return false;
+        }
+        n += recvd;
+    }
+
+    return true;
+}
+
+// this function uses perror & fprintf instead of the Log class, because it doesn't execute
+// as part of async-profiler (but instead, in fdtransfer).
+bool FdTransfer::serveRequests(pid_t pid) {
+    while (1) {
+        unsigned char request_buf[MAX_REQUEST_LENGTH];
+        struct fd_request *header = (struct fd_request *)request_buf;
+
+        bool eof;
+        if (!recv_all(_peer, header, sizeof(*header), &eof)) {
+            if (eof) {
+                // EOF means done
+                return true;
+            } else {
+                perror("recv()");
+                break;
+            }
+        }
+
+        if (header->length > MAX_REQUEST_LENGTH) {
+            fprintf(stderr, "too large request: %u\n", header->length);
+            break;
+        } else if (header->length < sizeof(*header)) {
+            fprintf(stderr, "bogus header length\n");
+            break;
+        }
+
+        const size_t left = header->length - sizeof(*header);
+        if (left > 0) {
+            if (!recv_all(_peer, &header[1], left, &eof)) {
+                if (eof) {
+                    fprintf(stderr, "truncated request\n");
+                    break;
+                } else {
+                    perror("recv()");
+                    break;
+                }
+            }
+        }
+
+        switch (header->type) {
+        case PERF_FD: {
+            struct perf_fd_request *request = (struct perf_fd_request*)header;
+            int perf_fd = -1;
+
+            uid_t _uid;
+            gid_t _gid;
+            pid_t _nspid;
+            pid_t target_pid = -1;
+            if (!get_process_info(request->tid, &_uid, &_gid, &_nspid, &target_pid)) {
```

Comment:
Can't we just check the existence of `/proc/[pid]/task/[tid]`?

Reply:
Or even better: `tgkill(pid, tid, 0)`

Reply:
I was trying too hard to use existing APIs :sweat_smile: also considered using `ThreadList` here. I agree that `/proc/pid/task/tid` is much simpler, or `tgkill`. Will use one of them

Reply:
Using `tgkill` to avoid the need for string formatting. I see it's available from Linux 2.5.75 and AP already uses it.